service/controller/resource/controlplanestatus/resource.go

```diff
@@ -0,0 +1,114 @@
+package controlplanestatus
+
+import (
+	"context"
+	"fmt"
+
+	infrastructurev1alpha2 "github.com/giantswarm/apiextensions/pkg/apis/infrastructure/v1alpha2"
+	"github.com/giantswarm/k8sclient/v3/pkg/k8sclient"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/giantswarm/operatorkit/controller/context/reconciliationcanceledcontext"
+	"github.com/giantswarm/operatorkit/controller/context/resourcecanceledcontext"
+
+	"github.com/giantswarm/cluster-operator/pkg/label"
+	"github.com/giantswarm/cluster-operator/service/controller/key"
+	"github.com/giantswarm/cluster-operator/service/internal/nodecount"
+)
+
+const (
+	Name = "controlplanestatus"
+)
+
+type Config struct {
+	K8sClient k8sclient.Interface
+	Logger    micrologger.Logger
+	NodeCount nodecount.Interface
+}
+
+type Resource struct {
+	k8sClient k8sclient.Interface
+	logger    micrologger.Logger
+	nodeCount nodecount.Interface
+}
+
+func New(config Config) (*Resource, error) {
+	if config.K8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.K8sClient must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+	if config.NodeCount == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.NodeCount must not be empty", config)
+	}
+
+	r := &Resource{
+		k8sClient: config.K8sClient,
+		logger:    config.Logger,
+		nodeCount: config.NodeCount,
+	}
+
+	return r, nil
+}
+
+func (r *Resource) Name() string {
+	return Name
+}
+
+func (r *Resource) ensure(ctx context.Context, obj interface{}) error {
+	cr := &infrastructurev1alpha2.G8sControlPlane{}
+	{
+		cp, err := key.ToG8sControlPlane(obj)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		err = r.k8sClient.CtrlClient().Get(ctx, key.ObjRefToNamespacedName(key.ObjRefFromG8sControlPlane(cp)), cr)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+	masterNodes, err := r.nodeCount.MasterCount(ctx, &cr)
+	if err != nil {
+		if nodecount.IsInterfaceError(err) {
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("not getting master nodes for tenant cluster %#q", key.ClusterID(cr)))
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+			resourcecanceledcontext.SetCanceled(ctx)
```

Comment:
This is the right thing to do here in the handler. It is only that we now also need to return at this point and not further execute the handler, because it can't.

Reply:
Done, I hope it was ok to return the error still after cancellation of the context.